### PR TITLE
fix: [PL-42600]: Calculate delegate cpu and memory usage from top command

### PR DESCRIPTION
**Pull request by jennyjames21**
## Describe your changes
Summary:
- Calculate CPU and memory usage from top command  for dynamic request handling of task functionality

Test
- Verified CPU and memory calculation and compare with top command

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56812)

### CODE REVIEW

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

File: 260-delegate/src/main/java/io/harness/delegate/service/LogPerformanceImpl.java [hunk 1]
```diff
@@ -109,4 +117,79 @@ private void logTopProcessesByCpuAndMemory() {
...
+      ProcessBuilder cpuProcessBuilder = new ProcessBuilder("top", "-b", "-n", "1", "-o", "%CPU");
+      cpuProcessBuilder.environment().put(TERM_ENV_VARIABLE, termEnv);
+      Process process = cpuProcessBuilder.start();
+
+      try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+        String line;
+        while ((line = reader.readLine()) != null) {
+          String[] processInfo = line.split("\\s+");
+          // Look for line, %Cpu(s):  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
+          if (isNotEmpty(processInfo[0]) && processInfo[0].equals("%Cpu(s):")) {
```
Comment (raghuAtWings) [inline 260-delegate/src/main/java/io/harness/delegate/service/LogPerformanceImpl.java:138]: should do case insensitive %Cpu

File: 260-delegate/src/main/java/io/harness/delegate/service/LogPerformanceImpl.java [hunk 2]
```diff
@@ -109,4 +117,79 @@ private void logTopProcessesByCpuAndMemory() {
...
+            List<String> cpuLine = new ArrayList<>();
+            cpuLine.addAll(Arrays.stream(processInfo).filter(EmptyPredicate::isNotEmpty).collect(Collectors.toList()));
+            if (isNotEmpty(cpuLine) && isNotEmpty(cpuLine.get(1)) && !resourceUsuageMap.containsKey("CPU")) {
```
Comment (raghuAtWings) [inline 260-delegate/src/main/java/io/harness/delegate/service/LogPerformanceImpl.java:141]: why double? can we not use float?

**Review by raghuAtWings — CHANGES_REQUESTED** [review]

File: 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java [hunk 3]
```diff
@@ -758,7 +758,12 @@ private String getDelegateConfig() {
   }
 
   private void maybeUpdateTaskRejectionStatus() {
-    final long currentRSSMB = MemoryHelper.getProcessMemoryMB();
+    Map<String, Double> resourceUsuageMap = logPerformanceImpl.getDelegateMemoryCpuUsage();
+    final double cpuUsage = resourceUsuageMap.get("CPU");
```
[the remark this answers is not in the document]
Comment (jennyjames21) [reply 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java:762]: add fallback logic to old flow

File: 260-delegate/src/main/java/io/harness/delegate/service/LogPerformanceImpl.java (hunk 1 above)
Comment (jennyjames21) [reply 260-delegate/src/main/java/io/harness/delegate/service/LogPerformanceImpl.java:138]: updated

File: 260-delegate/src/main/java/io/harness/delegate/service/LogPerformanceImpl.java (hunk 2 above)
Comment (jennyjames21) [reply 260-delegate/src/main/java/io/harness/delegate/service/LogPerformanceImpl.java:141]: calculation on the threshold and everywhere is double currently

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Review by raghuAtWings — APPROVED** [review]

**Comment by jennyjames21** [comment]
trigger messagecheck

**Comment by jennyjames21** [comment]
trigger messagecheck

**Pull request merged**